nuxeo-ai-core/src/main/java/org/nuxeo/ai/auto/AutoServiceImpl.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ai.metadata.AIMetadata;
+import org.nuxeo.ai.metadata.SuggestionMetadataAdapter;
+import org.nuxeo.ai.services.DocMetadataService;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.ecm.core.api.model.PropertyNotFoundException;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Autofill and AutoCorrect services
+ */
+public class AutoServiceImpl implements AutoService {
+
+    private static final Logger log = LogManager.getLogger(AutoServiceImpl.class);
+
+    @Override
+    public void calculateProperties(DocumentModel doc) {
+        calculateProperties(doc, ALL);
+    }
+
+    @Override
+    public void calculateProperties(DocumentModel doc, AUTO_ACTION action) {
+        SuggestionMetadataAdapter adapted = doc.getAdapter(SuggestionMetadataAdapter.class);
+        for (String xpath : adapted.getSuggestedProperties()) {
+            switch (action) {
+                case ALL:
+                    calculateAutoCorrect(adapted, xpath);
+                    calculateAutoFill(adapted, xpath);
+                    break;
+                case CORRECT:
+                    calculateAutoCorrect(adapted, xpath);
+                    break;
+                case FILL:
+                    calculateAutoFill(adapted, xpath);
+            }
+        }
+    }
+
+    protected void calculateAutoFill(SuggestionMetadataAdapter docMetadata, String xpath) {
+        float threshold = Framework.getService(ThresholdService.class)
+                                   .getAutoFillThreshold(docMetadata.getDoc(), xpath);
+        Property property;
+        boolean alreadyAutofilled = docMetadata.isAutoFilled(xpath);
+        try {
+            property = docMetadata.getDoc().getProperty(xpath);
+            if (!alreadyAutofilled && property.getValue() != null) {
+                // Can't set a non null value
+                log.debug("Unable to autofill property {} for doc {} because it has a value.", xpath,
+                          docMetadata.getDoc().getId());
+                return;
+            }
+        } catch (PropertyNotFoundException e) {
+            log.warn("Unknown autofill property {} ", xpath);
+            return;
+        }
+
+        AIMetadata.Label max = calculateMaxLabel(docMetadata.getSuggestionsByProperty(xpath), threshold);
+        if (max != null) {
+            String comment = String.format("Auto filled %s. (Confidence %s , Threshold %s)",
+                                           xpath, max.getConfidence(), threshold);
+            log.debug(comment);
+            property.setValue(max.getName());
+            Framework.getService(DocMetadataService.class)
+                     .updateAuto(docMetadata.getDoc(), AUTO_FILLED, xpath, null, comment);
+        } else {
+            if (alreadyAutofilled) {
+                // We autofilled but now the value didn't autofill so lets reset it
```

Comment:
Document in some place this behaviour, as this business logic is important to understand, and "accept".

Reply:
hmm, where's the best place for this?  I will add it to the readme